### ISSUE TITLE
 How to support other languages?

### ISSUE THREAD

**Issue by Chunhao-Li** [tomhosking/hrq-vae#5]
Is it possible to support other languages such as Chinese? What modifications do I need to make? Thank you.

**Comment by tomhosking** [comment]
Hi, this would be great to see!

It's possible, in theory - but a little difficult in practice. You would need to retrain the model, and you would need:
- training data in the target language, consisting of paraphrase pairs or clusters
- to tell the model to use a vocab/tokenizer in the target language. We use the huggingface `tokenizers` library so the model should support any tokenizer they provide
- a way of creating the "exemplars" used to guide training. We create exemplars based around syntactic chunks using `flair-nlp` but they only provide an English chunker. You could potentially use a POS tagger in another language, a constituency parser, or some other way of identifying sentences with similar syntactic structure.

**Comment by Chunhao-Li** [comment]
Thank you for your reply!
* I have a question about the `embeddings.pt` file in `pretrained-vocabs`. Where does it come from and what does it use for? 
* Also, how to save the trained model such that it can be used for inference?

**Comment by tomhosking** [comment]
The `embeddings.pt` is extracted from a pretrained model, eg:
```
from transformers import BertModel
import torch

model = BertModel.from_pretrained('bert-base-uncased')
bert_embeddings = model.embeddings.word_embeddings.weight.data

torch.save(bert_embeddings, '../data/pretrained-vocabs/bert-base-uncased.embeddings.pt')
```
I use it to initialise the token embeddings for the model.

The trained model should be saved by default to the run output folder - this is in `./runs/` by default. If you're running stuff from a notebook or script, you may need to pass in a value for `output_path` when you create the `Agent` object.

**Comment by Chunhao-Li** [comment]
For the `./runs/` folder, the predicted result of `dev` and `test` can be found, but the trained model does not exist here.

**Comment by tomhosking** [comment]
What code are you using to train the model? The usual training loop should create a file `./runs/<run_id>/model/checkpoint.pt` that contains the best performing model so far (as long as training goes on for longer than the `warmup_epochs` value).

**Comment by Chunhao-Li** [comment]
Thanks, it works. The problem was because that I set the `num_epochs` which was the same as `warm_epochs`. 
I've got some questions when I was trying to understand and modify the codes.
* What is `bottleneck` in the config file? Which part of the paper does it correspond to?
* I modified parts of the codes as you mentioned before to support other languages, but now the predicted output is always the empty string `''`. Do you have any ideas which part might be wrong?

**Comment by tomhosking** [comment]
OK great!

The `bottleneck` part of the config is mostly to do with the hierarchical quantization (Section 3.1). It maps the output of a transformer encoder to a hierarchy of discrete codes. There are also some settings to make sure that the semantic part of the encoding is passed to the decoder in the correct way.

Without knowing what changes you've made to the code, it's very difficult to know why it isn't working.

**Comment by Chunhao-Li** [comment]
Thanks for the response above. I've got some questions regarding to `generate_training_triples.py`. It seems to do more than descriptions in appendix C.
* What is the difference between `tags_to_preserve` and `NODES_TO_PRESERVE`? How did you choose those tags?
* The syntax in Chinese is a little bit different from English. What would happen if I use a Chinese POS tagger with almost different tags?

**Comment by tomhosking** [comment]
The script has the capability to do more than Appendix C, but we used the specific settings described in Appendix C:
```
python3 ./scripts/generate_training_triples.py  --use_diff_templ_for_sem --rate 1.0 --sample_size 26 --pos_templates --extended_stopwords --no_stopwords  --real_exemplars --exhaustive --template_dropout 0.3 --dataset mscoco-clusters --min_samples 0
```

`NODES_TO_PRESERVE` refers to constituency parses, whereas `tags_to_preserve` is used for templates based on chunking or POS tags. They were chosen to preserve "non semantic" tokens, like punctuation, prepositions and conjunctions.

I'm not familiar with Chinese syntax, but it should work OK - you will need to decide what parts of speech are primarily "syntactic" and which are more likely to contain semantic content. Then, preserve the words or tags that are mainly syntactic.

**Comment by Chunhao-Li** [comment]
Thanks. Regarding to the `bottleneck` mentioned above, I have some further questions about the config:
* Why there are two `modules` here ? And what is the `range` in the config?
* The `type` in modules are `vae` and `vqvae`, why not use `hrqvae`?

**Comment by tomhosking** [comment]
The input to the decoder is a combination of both the syntactic sketch and the semantic encoding - the `range` determines which heads of the encoder are used for which of these two types of encoding. The first part (heads 1 and 2) are a continuous, unpooled encoding (hence vae) - heads 6-8 are quantized, so the quantizer module is applied to them.

During development, I implemented HRQ-VAE as an extension of VQ-VAE. It's also possible to use the `hrqvae` module, but the config is slightly different:
```
        "modules": [
            {
                "range": [0,2],
                "type": "vae",
                "pooling": false
            },
            {
                "range": [2,8],
                "type": "hrqvae",
                "pooling": true,
                "quantizer": {
                    "codebook_size": 16,
                    "num_heads": 3,
                    "head_dropout": 0.1,
                    "head_dropout_keep_first": true,
                    "init_decay_weight": 0.5,

                    "init_scale": 0.9,
                    "init_embeds_xavier": false,
                    "init_embeds_truncnorm": false,
                    "init_embeds_uniform": true,
                    "gumbel_temp": 1.0,
                    "temp_schedule": true,
                    "temp_schedule_gamma": 10000,
                    "temp_min": 0.5,
                    "include_residual": false,
                    "kl_weight": 0.01,
                    "sqrt_distances": false,
                    "init_sphere": true,
                    "commitment_weight": 0,
                    "pre_norm": false,
                    "soft_gumbel": true
                }
            }
        ],
```

**Comment by Chunhao-Li** [comment]
Thank you. By the way, is it possible to use multiple gpus for training? What modifications do I need to make?

**Comment by tomhosking** [comment]
Unfortunately multiple GPUs aren't supported yet - it's really complicated to implement! If you're having issues fitting batches into memory, you can use gradient checkpointing by setting  `optim_batch_size` to a larger value than `batch_size` in the config.

**Comment by Chunhao-Li** [comment]
Hi, in `examples/tok.py`, does `pred_codes` act as the 'syntactic' input to the model. 
And what does the first `inference` do ? Could you explain `memory_eval` and `sep_encoding` a little bit?
``_, _, (_, _, _), memory_eval = agent.inference(
    data_loader.valid_loader,
    memory_keys_to_return=[f"sep_encoding_1{post_bottleneck}", f"sep_encoding_2{post_bottleneck}", "vq_codes"],
) ``

To generate multiple paraphrases, do you have any ideas on speeding up the inference process?

**Comment by tomhosking** [comment]
Yes, `pred_codes` is used as the syntactic sketch instead of the sketch corresponding to the input sentence.

The predicted sketch (or sketches) are sampled from the code predictor. The code predictor takes the semantic encoding as input. So, the first `inference` call is to generate the semantic encodings. Then it samples from the code predictor to get the predicted sketches `pred_codes`. Finally, there's another `inference` call to generate the output sentences from the original semantic encoding and the predicted syntactic encoding. `memory_eval` (and `memory` in the main torchseq code) is used as an all-purpose storage by the model, to make it easier to extract data from different parts of the model. `sep_encoding_{1,2}` are the semantic/syntactic encodings respectively, and `vq_codes` are the quantized representations of the syntactic encoding.

In terms of speeding up inference, it's a bit tricky to do in a single pass since it would require changing the batch size halfway through a forward pass (from `bsz` to `bsz * k`). But, the implementation of beam search in `code_predictor.infer()` is currently not batched so that's an obvious place to get some speedup. Depending on what GPU you have available you may also be able to increase the value of `eval_batch_size` for the two `inference` calls.

**Comment by Chunhao-Li** [comment]
Thank you. It finally works although there is room for improvement. It seems that training on a large dataset can improve significantly.
My question is that is it able to fine-tune the trained model ?
